File: app/src/main/java/com/kelsos/mbrc/ui/fragments/FeedbackFragment.java
package com.kelsos.mbrc.ui.fragments;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import com.kelsos.mbrc.R;

public class FeedbackFragment extends Fragment {

  @BindView(R.id.feedback_content)
  EditText feedbackEditText;

  public FeedbackFragment() {

  }

  @NonNull public static FeedbackFragment newInstance() {
    return new FeedbackFragment();
  }

  @Nullable
  @Override
  public View onCreateView(LayoutInflater inflater,
      @Nullable ViewGroup container,
      @Nullable Bundle savedInstanceState) {
    View view = inflater.inflate(R.layout.fragment_feedback, container, false);
    ButterKnife.bind(this, view);
    return view;
  }

  @OnClick(R.id.feedback_button)
  void onFeedbackButtonClicked() {
    final String feedbackText = feedbackEditText.getText().toString().trim();
    if (TextUtils.isEmpty(feedbackText)) {
      return;
    }

    Intent emailIntent = new Intent(Intent.ACTION_SEND);
    emailIntent.putExtra(Intent.EXTRA_EMAIL, new String[] { "kelsos@kelsos.net" });
    emailIntent.setType("message/rfc822");
    emailIntent.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.feedback_subject));
    emailIntent.putExtra(Intent.EXTRA_TEXT, feedbackText);
    startActivity(Intent.createChooser(emailIntent, getString(R.string.feedback_chooser_title)));
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/ui/help_feedback/FeedbackFragment.kt
package com.kelsos.mbrc.ui.help_feedback

import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.CheckBox
import android.widget.EditText
import androidx.core.content.FileProvider
import androidx.fragment.app.Fragment
import com.kelsos.mbrc.BuildConfig.APPLICATION_ID
import com.kelsos.mbrc.R
import com.kelsos.mbrc.logging.LogHelper
import com.kelsos.mbrc.utilities.RemoteUtils.getVersion
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import java.io.File

class FeedbackFragment : Fragment() {

  private lateinit var feedbackEditText: EditText
  private lateinit var deviceInfo: CheckBox
  private lateinit var logInfo: CheckBox
  private lateinit var feedbackButton: Button

  override fun onCreateView(
    inflater: LayoutInflater,
    container: ViewGroup?,
    savedInstanceState: Bundle?
  ): View? {
    val view = inflater.inflate(R.layout.fragment_feedback, container, false)
    feedbackEditText = view.findViewById(R.id.feedback_content);
    deviceInfo = view.findViewById(R.id.include_device_info)
    logInfo = view.findViewById(R.id.include_log_info)
    feedbackButton = view.findViewById(R.id.feedback_button)

    feedbackButton.setOnClickListener { onFeedbackButtonClicked() }

    LogHelper.logsExist(requireContext())
      .subscribeOn(Schedulers.io())
      .observeOn(AndroidSchedulers.mainThread())
      .subscribe({
        logInfo.isEnabled = true
      }) {

      }
    return view
  }

  private fun onFeedbackButtonClicked() {
    var feedbackText = feedbackEditText.text.toString().trim { it <= ' ' }
    if (TextUtils.isEmpty(feedbackText)) {
      return
    }

    feedbackButton.isEnabled = false

    if (deviceInfo.isChecked) {
      val device = Build.DEVICE
      val manufacturer = Build.MANUFACTURER
      val appVersion = requireContext().getVersion()
      val androidVersion = Build.VERSION.RELEASE

      feedbackText += getString(
        R.string.feedback_version_info,
        manufacturer,
        device,
        androidVersion,
        appVersion
      )
    }

    if (!logInfo.isChecked) {
      openChooser(feedbackText)
      return
    }

    LogHelper.zipLogs(requireContext())
      .subscribeOn(Schedulers.io())
      .observeOn(AndroidSchedulers.mainThread())
      .subscribe({
        openChooser(feedbackText, it)
      }) {
        openChooser(feedbackText)
      }
  }

  private fun openChooser(feedbackText: String, logs: File? = null) {
    val emailIntent = Intent(Intent.ACTION_SEND)
    emailIntent.putExtra(Intent.EXTRA_EMAIL, arrayOf("kelsos@kelsos.net"))
    emailIntent.type = "message/rfc822"
    emailIntent.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.feedback_subject))
    emailIntent.putExtra(Intent.EXTRA_TEXT, feedbackText)
    if (logs != null) {
      val logsUri = FileProvider.getUriForFile(requireContext(), "$APPLICATION_ID.fileprovider", logs)
      emailIntent.putExtra(Intent.EXTRA_STREAM, logsUri)
    }

    feedbackButton.isEnabled = true

    startActivity(Intent.createChooser(emailIntent, getString(R.string.feedback_chooser_title)))
  }

  companion object {
    fun newInstance(): FeedbackFragment {
      return FeedbackFragment()
    }
  }
}
